torch/nn/parallel/distributed.py

```diff
@@ -691,6 +728,52 @@ def __init__(
 
         self._setup_in_backward_optimizers()
 
+    def _register_delay_all_reduce_hook(
+        self,
+        bucket_cap_mb,
+        process_group,
+        param_to_hook_all_reduce,
+        device_ids,
+    ):
+        # 1. Create gradient buffer
+        device = torch.device("cpu") if device_ids is None else device_ids[0]
+        self._grad_buffer = torch.zeros(
+            sum([p.numel() for p in self._delay_all_reduce_params]),
+            device=device,
+        )
+
+        # 2. Broadcast the parameters
+        detached_params = [p.detach() for p in self._delay_all_reduce_params]
+        dist._broadcast_coalesced(process_group, detached_params, bucket_cap_mb, 0)
+
+        # 3. Hook all reduce to the specified parameter
+        world_size = dist.get_world_size(process_group)
+
+        def _delayed_all_reduce(grad):
+            self._grad_buffer.div_(world_size)
+            _ = dist.all_reduce(self._grad_buffer, group=process_group, async_op=True)
```

Comment:
where to wait for the all_reduce to complete? are you assuming optimizers will just wait for their grads?

Reply:
In our use case, we assume this happens before the final all reduce (DenseArch all reduce). When they are using the same PG, we don't have to explicitly wait for this all reduce. 

But I agree that this brings the risk of missing synchronization. Do you think I should comment here about our assumptions?